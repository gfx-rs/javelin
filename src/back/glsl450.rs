use crate::{
    back::glsl_common::{write_type, Error, StatementBuilder},
    Binding, BuiltIn, FastHashMap, Module, TypeInner,
};
use std::{
    fmt::{self},
    io::Write,
};

pub fn write(module: &Module, out: &mut impl Write) -> Result<(), Error> {
    writeln!(out, "#version 450 core")?;

    let mut counter = 0;
    let mut names = FastHashMap::default();

    let mut namer = |name: Option<&String>| {
        if let Some(name) = name {
            names.insert(name.clone(), ());
            name.clone()
        } else {
            counter += 1;
            while names.get(&format!("_{}", counter)).is_some() {
                counter += 1;
            }
            format!("_{}", counter)
        }
    };

    let mut structs = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let name = namer(ty.name.as_ref());
                let mut idxs = Vec::new();

                for member in members {
                    idxs.push(namer(member.name.as_ref()))
                }

                structs.insert(handle, (name, idxs));
            }
            _ => continue,
        }
    }

    // Do a second pass to build the structs
    // TODO: glsl is order dependent so we need to build structs in order
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let (name, idxs) = structs.get(&handle).unwrap();

                writeln!(out, "struct {} {{", name)?;
                for (member, name) in members.iter().zip(idxs.iter()) {
                    writeln!(
                        out,
                        "   {} {}",
                        write_type(member.ty, &module.types, &structs)?,
                        name
                    )?;
                }
                writeln!(out, "}};")?;
            }
            _ => continue,
        }
    }

    let mut globals_lookup = FastHashMap::default();

    for (handle, global) in module.global_variables.iter() {
        if let Some(Binding::BuiltIn(built_in)) = global.binding {
            let semantic = match built_in {
                BuiltIn::Position => "gl_position",
                BuiltIn::GlobalInvocationId => "gl_GlobalInvocationID",
                BuiltIn::BaseInstance => "gl_BaseInstance",
                BuiltIn::BaseVertex => "gl_BaseVertex",
                BuiltIn::ClipDistance => "gl_ClipDistance",
                BuiltIn::InstanceIndex => "gl_InstanceIndex",
                BuiltIn::VertexIndex => "gl_VertexIndex",
                BuiltIn::PointSize => "gl_PointSize",
                BuiltIn::FragCoord => "gl_FragCoord",
                BuiltIn::FrontFacing => "gl_FrontFacing",
                BuiltIn::SampleIndex => "gl_SampleID",
                BuiltIn::FragDepth => "gl_FragDepth",
                BuiltIn::LocalInvocationId => "gl_LocalInvocationID",
                BuiltIn::LocalInvocationIndex => "gl_LocalInvocationIndex",
                BuiltIn::WorkGroupId => "gl_WorkGroupID",
            };

            globals_lookup.insert(handle, String::from(semantic));
            continue;
        }

        if let Some(ref binding) = global.binding {
            write!(out, "layout({}) ", binding.writer())?;
        }

        let name = namer(global.name.as_ref());

        writeln!(
            out,
            "{}{} {};",
            global.class.write_glsl(),
            write_type(global.ty, &module.types, &structs)?,
            name
        )?;

        globals_lookup.insert(handle, name);
    }

    let mut functions = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, func) in module.functions.iter() {
        functions.insert(handle, namer(func.name.as_ref()));
    }

    let mut types = module.types.clone();
    let mut typifier = crate::proc::Typifier::new();

    // TODO: glsl is order dependent so we need to build functions in order
    for (handle, func) in module.functions.iter() {
        let name = functions.get(&handle).unwrap();

        writeln!(
            out,
            "{} {}({}) {{",
            func.return_type
                .map_or(Ok(String::from("void")), |ty| write_type(
                    ty, &types, &structs
                ))?,
            name,
            func.parameter_types
                .iter()
                .map(|ty| write_type(*ty, &types, &structs))
                .collect::<Result<Vec<_>, _>>()?
                .join(","),
        )?;

        let mut builder = StatementBuilder {
            functions: &functions,
            globals: &globals_lookup,
            locals_lookup: &func
                .local_variables
                .iter()
                .map(|(handle, local)| (handle, namer(local.name.as_ref())))
                .collect(),
            structs: &structs,
            args: &func
                .parameter_types
                .iter()
                .enumerate()
                .map(|(pos, _)| (pos as u32, namer(None)))
                .collect(),
            expressions: &func.expressions,
            types: &mut types,
            locals: &func.local_variables,
            typifier: &mut typifier,
        };

        for sta in func.body.iter() {
            writeln!(out, "{}", sta.write_glsl(module, &mut builder)?)?;
        }

        writeln!(out, "}}")?;
    }

    Ok(())
}

struct BindingWriter<'a> {
    inner: &'a Binding,
}

impl Binding {
    fn writer<'a>(&'a self) -> BindingWriter<'a> {
        BindingWriter { inner: self }
    }
}

impl<'a> fmt::Display for BindingWriter<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.inner {
            Binding::BuiltIn(_) => write!(f, ""), // Ignore because they are variables with a predefined name
            Binding::Location(location) => write!(f, "location={}", location),
            Binding::Descriptor { set, binding } => write!(f, "set={},binding={}", set, binding),
        }
    }
}
