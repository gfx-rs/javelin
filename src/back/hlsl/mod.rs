mod keywords;
mod writer;

use std::fmt::Error as FmtError;
use thiserror::Error;

pub use writer::Writer;

#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
pub struct ShaderModel(u16);

impl Default for ShaderModel {
    fn default() -> Self {
        ShaderModel(50)
    }
}

#[derive(Error, Debug)]
pub enum Error {
    #[error(transparent)]
    IoError(#[from] FmtError),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedShaderModel(ShaderModel),
    #[error("A scalar with an unsupported width was requested: {0:?} {1:?}")]
    UnsupportedScalar(crate::ScalarKind, crate::Bytes),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedBuiltIn(crate::BuiltIn),
    #[error("{0}")]
    Unimplemented(String), // TODO: Error used only during development
}

pub fn write_string(
    module: &crate::Module,
    info: &crate::valid::ModuleInfo,
    shader_model: ShaderModel,
) -> Result<String, Error> {
    let mut w = Writer::new(String::new(), shader_model);
    w.write(module, info)?;
    let output = w.finish();
    Ok(output)
}
