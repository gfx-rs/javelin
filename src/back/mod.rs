//! Functions which export shader modules into binary and text formats.

#[cfg(feature = "glsl-out")]
pub mod glsl;
#[cfg(feature = "msl-out")]
pub mod msl;
#[cfg(feature = "spv-out")]
pub mod spv;

impl crate::Expression {
    /// Returns the ref count, upon reaching which this expression
    /// should be considered for baking.
    #[allow(dead_code)]
    fn bake_ref_count(&self) -> usize {
        match *self {
            // accesses are never cached, only loads are
            crate::Expression::Access { .. } | crate::Expression::AccessIndex { .. } => !0,
            // image operations look better when isolated
            crate::Expression::ImageSample { .. } | crate::Expression::ImageLoad { .. } => 1,
            // cache expressions that are referenced multiple times
            _ => 2,
        }
    }
}
