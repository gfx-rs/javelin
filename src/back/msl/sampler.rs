#[derive(Clone, Debug, PartialEq)]
pub enum Coord {
    Normalized,
    Pixel,
}

impl Default for Coord {
    fn default() -> Self {
        Self::Normalized
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Address {
    Repeat,
    MirroredRepeat,
    ClampToEdge,
    ClampToZero,
    ClampToBorder,
}

impl Default for Address {
    fn default() -> Self {
        Self::ClampToEdge
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum BorderColor {
    TransparentBlack,
    OpaqueBlack,
    OpaqueWhite,
}

impl Default for BorderColor {
    fn default() -> Self {
        Self::TransparentBlack
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Filter {
    Nearest,
    Linear,
}

impl Default for Filter {
    fn default() -> Self {
        Self::Nearest
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum CompareFunc {
    Never,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    Always,
}

impl Default for CompareFunc {
    fn default() -> Self {
        Self::Never
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct InlineSampler {
    pub coord: Coord,
    pub address: [Address; 3],
    pub border_color: BorderColor,
    pub mag_filter: Filter,
    pub min_filter: Filter,
    pub mip_filter: Option<Filter>,
    pub compare_func: CompareFunc,
}
