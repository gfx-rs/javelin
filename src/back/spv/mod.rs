mod layout;
pub mod parser;

use spirv::*;

struct Module {
    physical_layout: PhysicalLayout,
    logical_layout: LogicalLayout,
}

impl Module {
    fn new(header: &crate::Header) -> Self {
        Module {
            physical_layout: PhysicalLayout::new(header, 0x0u32),
            logical_layout: LogicalLayout::new(),
        }
    }
}

struct PhysicalLayout {
    magic_number: Word,
    version: Word,
    generator: Word,
    bound: Word,
    instruction_schema: Word,
}

impl PhysicalLayout {
    fn new(header: &crate::Header, instruction_schema: Word) -> Self {
        let version: Word = (0x0u32 << 24)
            | ((header.version.0 as u32) << 16)
            | ((header.version.1 as u32) << 8)
            | header.version.2 as u32;

        PhysicalLayout {
            magic_number: MAGIC_NUMBER,
            version,
            generator: header.generator,
            bound: 0,
            instruction_schema,
        }
    }

    fn in_words(&self) -> Vec<Word> {
        let mut words: Vec<Word> = vec![];

        words.push(self.magic_number);
        words.push(self.version);
        words.push(self.generator);

        match self.bound {
            0 => panic!("Bound cannot be 0"),
            _ => words.push(self.bound),
        }

        words.push(self.instruction_schema);
        words
    }
}

pub(crate) struct LogicalLayout {
    capabilities: Vec<Word>,
    extensions: Vec<Word>,
    ext_inst_imports: Vec<Word>,
    memory_model: Vec<Word>,
    entry_points: Vec<Word>,
    execution_modes: Vec<Word>,
    debugs: Vec<Word>,
    annotations: Vec<Word>,
    type_declarations: Vec<Word>,
    constants: Vec<Word>,
    global_variables: Vec<Word>,
    function_declarations: Vec<Word>,
    function_definitions: Vec<Word>,
}

impl LogicalLayout {
    fn new() -> Self {
        LogicalLayout {
            capabilities: vec![],
            extensions: vec![],
            ext_inst_imports: vec![],
            memory_model: vec![],
            entry_points: vec![],
            execution_modes: vec![],
            debugs: vec![],
            annotations: vec![],
            type_declarations: vec![],
            constants: vec![],
            global_variables: vec![],
            function_declarations: vec![],
            function_definitions: vec![],
        }
    }

    fn in_words(&self) -> Vec<u32> {
        let mut words: Vec<Word> = vec![];

        for capability in self.capabilities.iter() {
            words.push(*capability);
        }

        for extension in self.extensions.iter() {
            words.push(*extension);
        }

        for ext_inst_import in self.ext_inst_imports.iter() {
            words.push(*ext_inst_import);
        }

        for memory_model in self.memory_model.iter() {
            words.push(*memory_model);
        }

        for entry_point in self.entry_points.iter() {
            words.push(*entry_point);
        }

        for execution_mode in self.execution_modes.iter() {
            words.push(*execution_mode);
        }

        for debug in self.debugs.iter() {
            words.push(*debug);
        }

        for annotation in self.annotations.iter() {
            words.push(*annotation);
        }

        for type_declaration in self.type_declarations.iter() {
            words.push(*type_declaration);
        }

        for constants in self.constants.iter() {
            words.push(*constants);
        }

        for global_variables in self.global_variables.iter() {
            words.push(*global_variables);
        }

        for function_declaration in self.function_declarations.iter() {
            words.push(*function_declaration);
        }

        for function_definition in self.function_definitions.iter() {
            words.push(*function_definition);
        }

        words
    }
}

struct Instruction {
    op: Op,
    wc: u32,
    type_id: Option<Word>,
    result_id: Option<Word>,
    operands: Vec<Word>,
}

impl Instruction {
    fn new(op: Op) -> Self {
        Instruction {
            op,
            wc: 1, // Always start at 1 for the first word (OP + WC),
            type_id: None,
            result_id: None,
            operands: vec![],
        }
    }

    fn set_type(&mut self, id: Word) {
        assert!(self.type_id.is_none(), "Type can only be set once");
        self.type_id = Some(id);
        self.wc += 1;
    }

    fn set_result(&mut self, id: Word) {
        assert!(self.result_id.is_none(), "Result can only be set once");
        self.result_id = Some(id);
        self.wc += 1;
    }

    fn add_operand(&mut self, operand: Word) {
        self.operands.push(operand);
        self.wc += 1;
    }

    fn add_operands(&mut self, operands: Vec<Word>) {
        for operand in operands {
            self.add_operand(operand);
        }
    }

    fn to_words(&self) -> Vec<Word> {
        let mut words = Vec::with_capacity(self.wc as usize);

        let wc_op = (self.wc << 16 | self.op as u32) as u32;

        words.push(wc_op);

        if self.type_id.is_some() {
            words.push(self.type_id.unwrap());
        }

        if self.result_id.is_some() {
            words.push(self.result_id.unwrap());
        }

        for operand in self.operands.iter() {
            words.push(*operand);
        }

        words
    }
}
