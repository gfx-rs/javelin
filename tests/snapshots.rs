use serde::Deserialize;

#[derive(Hash, PartialEq, Eq, Deserialize)]
enum Stage {
    Vertex,
    Fragment,
    Compute,
}

#[derive(Hash, PartialEq, Eq, Deserialize)]
struct BindSource {
    stage: Stage,
    group: u32,
    binding: u32,
}

#[derive(Deserialize)]
struct BindTarget {
    #[serde(default)]
    buffer: Option<u8>,
    #[serde(default)]
    texture: Option<u8>,
    #[serde(default)]
    sampler: Option<u8>,
    #[serde(default)]
    mutable: bool,
}

#[derive(Default, Deserialize)]
struct Parameters {
    #[serde(default)]
    spv_flow_dump_prefix: String,
    #[cfg_attr(not(feature = "spv-out"), allow(dead_code))]
    spv_capabilities: naga::FastHashSet<spirv::Capability>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    mtl_bindings: naga::FastHashMap<BindSource, BindTarget>,
}

#[cfg(feature = "msl-out")]
fn check_output_msl(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::msl;
    let mut binding_map = msl::BindingMap::default();
    for (key, value) in params.mtl_bindings.iter() {
        binding_map.insert(
            msl::BindSource {
                stage: match key.stage {
                    Stage::Vertex => naga::ShaderStage::Vertex,
                    Stage::Fragment => naga::ShaderStage::Fragment,
                    Stage::Compute => naga::ShaderStage::Compute,
                },
                group: key.group,
                binding: key.binding,
            },
            msl::BindTarget {
                buffer: value.buffer,
                texture: value.texture,
                sampler: value.sampler,
                mutable: value.mutable,
            },
        );
    }
    let options = msl::Options {
        lang_version: (1, 0),
        spirv_cross_compatibility: false,
        binding_map,
    };
    let (msl, _) = msl::write_string(&module, &options).unwrap();
    insta::assert_snapshot!(format!("{}{}", name, ".msl"), msl);
}

#[cfg(feature = "spv-out")]
fn check_output_spv(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::spv;
    use rspirv::binary::Disassemble;
    let spv = spv::write_vec(
        &module,
        spv::WriterFlags::NONE,
        params.spv_capabilities.clone(),
    )
    .unwrap();
    let dis = rspirv::dr::load_words(spv)
        .expect("Produced invalid SPIR-V")
        .disassemble();
    insta::assert_snapshot!(format!("{}{}", name, ".spvasm"), dis);
}

#[cfg(feature = "wgsl-in")]
fn convert_wgsl(name: &str) {
    let params =
        match std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".param.ron")) {
            Ok(string) => ron::de::from_str(&string).expect("Couldn't find param file"),
            Err(_) => Parameters::default(),
        };

    let module = naga::front::wgsl::parse_str(
        &std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".wgsl"))
            .expect("Couldn't find wgsl file"),
    )
    .unwrap();
    naga::proc::Validator::new().validate(&module).unwrap();

    #[cfg(feature = "msl-out")]
    check_output_msl(&module, name, &params);
    #[cfg(feature = "spv-out")]
    check_output_spv(&module, name, &params);
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_quad() {
    convert_wgsl("quad");
}
