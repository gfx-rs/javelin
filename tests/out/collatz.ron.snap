---
source: tests/snapshots.rs
expression: output
---
(
    types: [
        (
            name: None,
            inner: Vector(
                size: Tri,
                kind: Uint,
                width: 4,
            ),
        ),
        (
            name: None,
            inner: Scalar(
                kind: Uint,
                width: 4,
            ),
        ),
        (
            name: None,
            inner: Array(
                base: 2,
                size: Dynamic,
                stride: Some(4),
            ),
        ),
        (
            name: Some("PrimeIndices"),
            inner: Struct(
                block: true,
                members: [
                    (
                        name: Some("data"),
                        span: None,
                        ty: 3,
                    ),
                ],
            ),
        ),
    ],
    constants: [
        (
            name: None,
            specialization: None,
            inner: Scalar(
                width: 4,
                value: Uint(0),
            ),
        ),
        (
            name: None,
            specialization: None,
            inner: Scalar(
                width: 4,
                value: Uint(1),
            ),
        ),
        (
            name: None,
            specialization: None,
            inner: Scalar(
                width: 4,
                value: Uint(2),
            ),
        ),
        (
            name: None,
            specialization: None,
            inner: Scalar(
                width: 4,
                value: Uint(3),
            ),
        ),
    ],
    global_variables: [
        (
            name: Some("global_id"),
            class: Input,
            binding: Some(BuiltIn(GlobalInvocationId)),
            ty: 1,
            init: None,
            interpolation: None,
            storage_access: (
                bits: 0,
            ),
        ),
        (
            name: Some("v_indices"),
            class: Storage,
            binding: Some(Resource(
                group: 0,
                binding: 0,
            )),
            ty: 4,
            init: None,
            interpolation: None,
            storage_access: (
                bits: 3,
            ),
        ),
    ],
    functions: [
        (
            name: Some("collatz_iterations"),
            arguments: [
                (
                    name: Some("n_base"),
                    ty: 2,
                ),
            ],
            return_type: Some(2),
            local_variables: [
                (
                    name: Some("n"),
                    ty: 2,
                    init: None,
                ),
                (
                    name: Some("i"),
                    ty: 2,
                    init: Some(1),
                ),
            ],
            expressions: [
                GlobalVariable(1),
                GlobalVariable(2),
                FunctionArgument(0),
                LocalVariable(1),
                Constant(1),
                LocalVariable(2),
                Load(
                    pointer: 4,
                ),
                Constant(2),
                Binary(
                    op: LessEqual,
                    left: 7,
                    right: 8,
                ),
                Load(
                    pointer: 4,
                ),
                Constant(3),
                Binary(
                    op: Modulo,
                    left: 10,
                    right: 11,
                ),
                Constant(1),
                Binary(
                    op: Equal,
                    left: 12,
                    right: 13,
                ),
                Load(
                    pointer: 4,
                ),
                Constant(3),
                Binary(
                    op: Divide,
                    left: 15,
                    right: 16,
                ),
                Constant(4),
                Load(
                    pointer: 4,
                ),
                Binary(
                    op: Multiply,
                    left: 18,
                    right: 19,
                ),
                Constant(2),
                Binary(
                    op: Add,
                    left: 20,
                    right: 21,
                ),
                Load(
                    pointer: 6,
                ),
                Constant(2),
                Binary(
                    op: Add,
                    left: 23,
                    right: 24,
                ),
                Load(
                    pointer: 6,
                ),
            ],
            body: [
                Store(
                    pointer: 4,
                    value: 3,
                ),
                Loop(
                    body: [
                        Emit((
                            start: 6,
                            end: 7,
                        )),
                        Emit((
                            start: 8,
                            end: 9,
                        )),
                        If(
                            condition: 9,
                            accept: [
                                Break,
                            ],
                            reject: [],
                        ),
                        Emit((
                            start: 9,
                            end: 10,
                        )),
                        Emit((
                            start: 11,
                            end: 12,
                        )),
                        Emit((
                            start: 13,
                            end: 14,
                        )),
                        If(
                            condition: 14,
                            accept: [
                                Emit((
                                    start: 14,
                                    end: 15,
                                )),
                                Emit((
                                    start: 16,
                                    end: 17,
                                )),
                                Store(
                                    pointer: 4,
                                    value: 17,
                                ),
                            ],
                            reject: [
                                Emit((
                                    start: 18,
                                    end: 20,
                                )),
                                Emit((
                                    start: 21,
                                    end: 22,
                                )),
                                Store(
                                    pointer: 4,
                                    value: 22,
                                ),
                            ],
                        ),
                        Emit((
                            start: 22,
                            end: 23,
                        )),
                        Emit((
                            start: 24,
                            end: 25,
                        )),
                        Store(
                            pointer: 6,
                            value: 25,
                        ),
                    ],
                    continuing: [],
                ),
                Emit((
                    start: 25,
                    end: 26,
                )),
                Return(
                    value: Some(26),
                ),
            ],
        ),
    ],
    entry_points: {
        (Compute, "main"): (
            early_depth_test: None,
            workgroup_size: (1, 1, 1),
            function: (
                name: Some("main"),
                arguments: [],
                return_type: None,
                local_variables: [],
                expressions: [
                    GlobalVariable(1),
                    GlobalVariable(2),
                    AccessIndex(
                        base: 2,
                        index: 0,
                    ),
                    Load(
                        pointer: 1,
                    ),
                    AccessIndex(
                        base: 4,
                        index: 0,
                    ),
                    Access(
                        base: 3,
                        index: 5,
                    ),
                    AccessIndex(
                        base: 2,
                        index: 0,
                    ),
                    Load(
                        pointer: 1,
                    ),
                    AccessIndex(
                        base: 8,
                        index: 0,
                    ),
                    Access(
                        base: 7,
                        index: 9,
                    ),
                    Load(
                        pointer: 10,
                    ),
                    Call(1),
                ],
                body: [
                    Emit((
                        start: 2,
                        end: 11,
                    )),
                    Call(
                        function: 1,
                        arguments: [
                            11,
                        ],
                        result: Some(12),
                    ),
                    Store(
                        pointer: 6,
                        value: 12,
                    ),
                    Return(
                        value: None,
                    ),
                ],
            ),
        ),
    },
)
